samples/bluetooth/hci_spi/src/main.c

```diff
@@ -0,0 +1,295 @@
+#include <bluetooth/hci.h>
+#include <bluetooth/buf.h>
+#include <bluetooth/hci_raw.h>
+
+#include "common/log.h"
+
+#define HCI_CMD                 0x01
+#define HCI_ACL                 0x02
+#define HCI_SCO                 0x03
+#define HCI_EVT                 0x04
+
+/* Special Values */
+#define SPI_WRITE               0x0A
+#define SPI_READ                0x0B
+#define READY_NOW               0x02
+#define SANITY_CHECK            0x02
+
+/* Offsets */
+#define STATUS_HEADER_READY     0
+#define STATUS_HEADER_TOREAD    3
+
+#define PACKET_TYPE             0
+#define EVT_BLUE_INITIALIZED    0x01
+
+#define SPI_HCI_DEV_NAME        CONFIG_BT_CONTROLLER_TO_HOST_SPI_DEV_NAME
+#define GPIO_IRQ_DEV_NAME       CONFIG_BT_CONTROLLER_TO_HOST_SPI_IRQ_DEV_NAME
+#define GPIO_IRQ_PIN            CONFIG_BT_CONTROLLER_TO_HOST_SPI_IRQ_PIN
+
+/* Needs to be aligned with the SPI master buffer size */
+#define SPI_MAX_MSG_LEN         255
+
+static u8_t rxmsg[SPI_MAX_MSG_LEN];
+static u8_t txmsg[SPI_MAX_MSG_LEN];
+
+/* HCI buffer pools */
+#define CMD_BUF_SIZE BT_BUF_RX_SIZE
+
+NET_BUF_POOL_DEFINE(cmd_tx_pool, CONFIG_BT_HCI_CMD_COUNT, CMD_BUF_SIZE,
+		    BT_BUF_USER_DATA_MIN, NULL);
+
+#if defined(CONFIG_BT_CONTROLLER)
+#define BT_L2CAP_MTU (CONFIG_BT_CONTROLLER_TX_BUFFER_SIZE - \
+		      BT_L2CAP_HDR_SIZE)
+#else
+#define BT_L2CAP_MTU 65 /* 64-byte public key + opcode */
+#endif /* CONFIG_BT_CONTROLLER */
+
+/* Data size needed for ACL buffers */
+#define BT_BUF_ACL_SIZE BT_L2CAP_BUF_SIZE(BT_L2CAP_MTU)
+
+#if defined(CONFIG_BT_CONTROLLER_TX_BUFFERS)
+#define TX_BUF_COUNT CONFIG_BT_CONTROLLER_TX_BUFFERS
+#else
+#define TX_BUF_COUNT 6
+#endif
+
+NET_BUF_POOL_DEFINE(acl_tx_pool, TX_BUF_COUNT, BT_BUF_ACL_SIZE,
+		    BT_BUF_USER_DATA_MIN, NULL);
+
+static struct device *spi_hci_dev;
+static struct device *gpio_dev;
+static BT_STACK_NOINIT(bt_tx_thread_stack, CONFIG_BT_HCI_TX_STACK_SIZE);
+static struct k_thread bt_tx_thread_data;
+
+static K_SEM_DEFINE(sem_spi_rx, 0, 1);
+static K_SEM_DEFINE(sem_spi_tx, 0, 1);
+
+static inline int spi_send(struct net_buf *buf)
+{
+	u8_t header_master[5] = { 0 };
+	u8_t header_slave[5] = { READY_NOW, SANITY_CHECK,
+				 0x00, 0x00, 0x00 };
+	u8_t ret;
+
+	SYS_LOG_DBG("buf %p type %u len %u", buf, bt_buf_get_type(buf),
+		    buf->len);
+
+	switch (bt_buf_get_type(buf)) {
+	case BT_BUF_ACL_IN:
+		net_buf_push_u8(buf, HCI_ACL);
+		break;
+	case BT_BUF_EVT:
+		net_buf_push_u8(buf, HCI_EVT);
+		break;
+	default:
+		SYS_LOG_ERR("Unknown type %u", bt_buf_get_type(buf));
+		net_buf_unref(buf);
+		return -EINVAL;
+	}
+
+	if (buf->len > SPI_MAX_MSG_LEN) {
+		SYS_LOG_ERR("TX message too long");
+		net_buf_unref(buf);
+		return -EINVAL;
+	}
+	header_slave[STATUS_HEADER_TOREAD] = buf->len;
+
+	gpio_pin_write(gpio_dev, GPIO_IRQ_PIN, 1);
+
+	/* Coordinate transfer lock with the spi rx thread */
+	k_sem_take(&sem_spi_tx, K_FOREVER);
+	do {
+		ret = spi_transceive(spi_hci_dev, header_slave, 5,
+				     header_master, 5);
+		if (ret < 0) {
```

Comment:
Nice catch. Will fix this here, below, and recompile with -Wall -Wextra to make sure nothing else like this is around in this app.